numerics/src/tools/NumericsMatrix.h

```diff
@@ -66,6 +70,9 @@ struct NumericsMatrix
 
 };
 
+/*! RawNumericsMatrix is used without conversion in python */
+typedef NumericsMatrix RawNumericsMatrix;
+
 /*! Available types of storage for NumericsMatrix */
```

Comment:
NumericsMatrix is converted through swig typemaps, this is nice but not in all situations. Also as swig typemaps are totally opaque for the user (what they do do not appear clearly in the documentation) sometimes it is difficult to infer what is done. I had great difficulties to understand when and why my NumericsMatrix was reallocated and why MPI communicator and MUMPS id were lost.  For example, in the master, when you are writing NM_create(1, 1, 1) in python, it is converted to something else on the fly!

```python
In [3]: NM_create(1,1,1)
Out[3]: <siconos.numerics.SparseBlockStructuredMatrix; proxy of <Swig Object of type 'struct SparseBlockStructuredMatrix *' at 0x7f804b1c69f0> >

In [4]: NM_create(0,1,1)
Out[4]: array([[0.]])
```

You do not get a NumericsMatrix but either a python object (type 0, type 2) or a SparseBlockStructuredMatrix! (type 1). The documentation is wrong, and NM_create is totally useless.

Swig typemaps will kill us. 

So, I found that solution is not so bad: we keep NumericsMatrix as a convertible type for python and we add a RawNumericsMatrix that is never changed in typemaps.  The whole NM_* API becomes usable from python. With this patch, I tried to fix all returns to NumericxMatrix so they are usable. I added a NM_convert that pass the NumericsMatrix to the typemaps stuff (in the case type 1, the result is weird, do we really, really need this ?). 

```python
In [5]: NM_create(1,1,1)
Out[5]: <siconos.numerics.NumericsMatrix; proxy of <Swig Object of type 'RawNumericsMatrix *' at 0x7f401d512f00> >

In [6]: NM_convert(NM_create(1,1,1))
Out[6]: <siconos.numerics.SparseBlockStructuredMatrix; proxy of <Swig Object of type 'struct SparseBlockStructuredMatrix *' at 0x7f401d512b40> >
```

I understand that it has nothing to do in the C alone part of numerics, but I think that this does not hurt so much and the benefit is that the documentation for type conversion can be made visible for everybody and can be done at the same place as the other documentation: we know what should happen.

Also, following this idea, some other specific functions as NM_convert exist in the swig files, I would suggest to move them in the sources in order to make them visible.

If we follow the idea that the source code should know nothing about bindings, moving this in the swig files means to adapt sphinx/doxygen to parse the swig files (is it possible ?) or to deal with an external part in C that would be associated to the swig stuff. Then it is not clear how to distinguish between convertible returns and non convertible ones. It is some work also and I may not have the time to do that quickly. Maybe we can raise an issue to think about it ?

Reply:
First thanks for looking into that. I think you must be one of the first to try to NM_create in python. I understand the frustration with the opacity of SWIG. I replied to a few of your points. I think it could be also a good idea to talk over the phone of via video.

For me the technical solution of using a typemap is C just to avoid some SWIG behaviour is not good. The C file should be influenced by the bindings. It would create 2 types visible to the user (in code and the documentation). And the only reason would be not to trigger a SWIG typemap. If I was to look into a software that would have this kind of behaviour, I would have a very very hard time to make my head work around this. It would make me inclined to not consider it further. Even for myself, I am not sure I will remember that new type and its meaning in 6 months.

I could see 2 approaches:
- disable the typemap for selected functions in NumericsMatrix.h. See http://www.swig.org/Doc3.0/SWIGDocumentation.html#Typemaps_nn12 and the sections on deleting a typemap. We can then remove the typemaps, do the ```%include NumericsMatrix.h```, and then redefine the typemaps.
- remove the "else" part of the "typemap out" on NumericsMatrix. When I look at functions returning NumericsMatrix (via ```git grep 'NumericsMatrix *\*[^=]*(.*)'``` ), I think it matches pretty well the ones where you use RawNumericsMatrix.

I think the biggest issue here is that the bindings behaviour was never really specified. I worked on those SWIG bindings quite a bit, but just to make the already "defined" behaviour work. I was quite frustrated that some conversion code was there, but was not working at all.
Now things are working better, but maybe not in the way we want. Rather than rushing a commit, I think it would be good to sit down and specify a bit the behaviour we want.

I just thought that maybe we don't want to convert all the NumericsMatrix to python objects. We can also augment NumericsMatrix so has to have a "array"-like behaviour (with more ```__setitem__``` and ```__getitem__``` methods).

We also have the long-standing issue that Siconos(Vector|Matrix) can't use non-managed memory. Therefore, modifying the python matrix would not change the Siconos object. This behaviour should also be documented.

Reply:
@xhub Nice review and very interesting comments. As you said " I think it would be good to sit down and specify a bit the behaviour we want", I agree with that. 

I propose to accept the PR for the following reasons: 
- we need it to evaluate the interest of MPI and MUMPS for our quite small problems and I prefer to merge before the master will diverge with coming commits on refactoring of the cmake files.
- The problem of swig and the hazardous typemaps is a problem that we encounter at any expert use of the python interface. For me, manipulations of data and directors are barely unusable, and it renders the interface very opaque and very unstable. I already had some code in order to simplify the swig/python behavior. This is not comfortable. If I was an external users,  after 5 minutes, I would switch to C++ interface, or use another software :-) 

In clear, the python interface cannot be used in expert mode. This is the current situation since 2 years. We cannot fix it quickly in this PR concerning MUMPS.

We have to open an issue that will yield a thorough refactoring of swig wrappers.

@xhub could you open an issue after the merge ?